### PR TITLE
fix(#199): Auto-tag react fragments

### PR DESCRIPTION
**Pull request by roycrippen4**
Auto-closes react fragment tags in `js`, `jsx`, and `tsx` files.

### CODE REVIEW

**Comment by PriceHiller** [comment]
Assigning myself as a reminder to look at this in a bit when I'm available.

**Comment by roycrippen4** [comment]
@PriceHiller No problem. I'm writing a few more tests to ensure it's working correctly. 

I know there's a possible issue in `.js` files if there isn't any `jsx` specific syntax to match against. If you just open a new `.js` file and the first `jsx` syntax is a react fragment it won't work. I don't think this is necessarily a bad thing as it'll be almost impossible to handle such a case.

**Comment by roycrippen4** [comment]
I found a couple issues when running the tests. I'm working on a fix now.

**Comment by PriceHiller** [comment]
If the tests fail due to incompatibility between this and a less important feature (like automatically adding a closing tag for an opening tag without a closing partner) and it proves to be too difficult to resolve the compatilibty between them just list out the compatability issues and we can decide if it's worth losing something for this.

Fragment closing seems to be quite popular in terms of requested features so I'm ok with losing lesser known/used features for it.

If that turns out to be the case though, I'll be marking this as the start of autotag 1.0.0 and yanking some things that have been marked as deprecated for a while.

**Comment by roycrippen4** [comment]
It should all be compatible. I just made an error with boolean logic and need one extra check function and (hopefully) everything should work. There are a couple of tests failing on the main branch. I'm making my measure of success is the same tests fail on this PR as they do on main.

**Comment by roycrippen4** [comment]
I'm pretty sure those are the same tests that are already failing on the main branch. Are those tests related to #190 and #196?

**Comment by PriceHiller** [comment]
> I'm pretty sure those are the same tests that are already failing on the main branch. Are those tests related to #190 and #196?

Yeah, those tests appear to be failing due to an update upstream in `nvim-treesitter` for the js/ts parsers.

Probably not your fault based on what I'm seeing locally. Entirely separate issue (though it is a problem).

**Comment by roycrippen4** [comment]
We should make an issue to get them fixed so at least it's documented.

**Comment by PriceHiller** [comment]
> We should make an issue to get them fixed so at least it's documented.

Already done 😉 (#205)

**Comment by PriceHiller** [comment]
Ping me when you're good to go over here (or convert this PR to a draft and then promote it when you're ready) and I'll take a look

**Comment by roycrippen4** [comment]
It's good to go

**Comment by roycrippen4** [comment]
I mentioned it #199, but this PR doesn't fix #185.

File: lua/nvim-ts-autotag/utils.lua [hunk 1]
```diff
@@ -2,6 +2,68 @@ local log = require("nvim-ts-autotag._log")
...
+---@return boolean
+function M.is_react_file()
+    local ft = vim.bo.ft
+    -- check filetypes first.
+    if ft == "javascriptreact" or ft == "typescriptreact" then
+        return true
+    elseif ft ~= "javascript" then
+        return false
+    end
+
```
Comment (PriceHiller) [inline lua/nvim-ts-autotag/utils.lua:14]: Just a simple comment to add, as the logic around the checks made me scratch my head for a sec before I realized *why* you were doing that. (Do correct the suggestion if I'm wrong).

```suggestion
-- If we are in a `javascript` file, then check the content to see if the
-- current file counts as a react file
```

**Review by PriceHiller — CHANGES_REQUESTED** [review]
Overall, it looks great!

Can you run `make format` for me and push those changes?You'll need `stylua` installed for the formatting.

I really need to add a CI check for formattingn now that I've noticed this 😅

File: lua/nvim-ts-autotag/utils.lua (hunk 1 above)
Comment (roycrippen4) [reply lua/nvim-ts-autotag/utils.lua:14]: That's precisely the problem that function is trying to solve, you nailed it. The low-hanging-fruit is the case where we are in a `jsx` or `tsx` file, or not in a `js` file. We need logic to determine if react is being used if we are in a `js` file. Javascript. I love it. lol

**Comment by PriceHiller** [comment]
Also, for clarification's sake, do I need to close #185 once this is merged?

**Comment by roycrippen4** [comment]
No, #185 will get it's own PR. I probably won't get it done today, but in the near future.

**Comment by PriceHiller** [comment]
> No, #185 will get it's own PR. I probably won't get it done today, but in the near future.

Gotcha, no rush on it. You're doing god's work right now. I'll leave that issue alone then.

Just hit it with a `make format` and I think we're good to go.

**Comment by PriceHiller** [comment]
Awesome, thanks again @roycrippen4 -- great work!

**Comment by roycrippen4** [comment]
Happy to help!

**Comment by abilkhan024** [comment]
Is there a way to opt out from this feature, it causes problems when i want to pass type arguments by adding </>. E.g.

**Comment by abilkhan024** [comment]
Is there a way to opt out from this feature, or potentially change current behavior. Right now it causes problems when i want to pass type arguments by adding </>. For example:
typing: useState<>
results: useState<></>
I didn't find work around for that, pretty sure that was not intentional behaviour.

**Pull request merged**